From ab6f62de6a597c78ba9cdb679d442a9b7996df08 Mon Sep 17 00:00:00 2001
From: WeiguangTWK <weiguangtwk@outlook.com>
Date: Wed, 1 Jul 2026 15:27:27 +0800
Subject: [PATCH] Neko LineageOS: Introduce SELinux infra

We achieve many of environment hide using SELinux policy like deny untrusted_app & isolated_app to reach lineage files
therefore many new types are created and most of major Neko-LineageOS features are build upon this, making it a impor-
-tant foundation for following feature patches

Change-Id: I325fe0c5e6f59c67a7ce16a3fc2057e10eb6b1bc
---
 common/private/file.te                       |  4 +-
 common/private/file_contexts                 | 46 ++++++++++++++++++++
 common/private/init.te                       |  2 +
 common/private/lineage_platform_framework.te |  2 +
 common/private/lineage_privapp_pkg.te        |  9 ++++
 common/private/lineage_visibility_access.te  |  3 ++
 common/private/system_server.te              |  6 ++-
 common/private/untrusted_app.te              | 20 ++++++++-
 common/vendor/file_contexts                  |  4 ++
 common/vendor/hal_lineage_health_default.te  |  2 +
 qcom/vendor/hal_perf_default.te              |  1 +
 11 files changed, 96 insertions(+), 3 deletions(-)
 create mode 100644 common/private/init.te
 create mode 100644 common/private/lineage_platform_framework.te
 create mode 100644 common/private/lineage_privapp_pkg.te
 create mode 100644 common/private/lineage_visibility_access.te

diff --git a/common/private/file.te b/common/private/file.te
index 6fb480b..d359cef 100644
--- a/common/private/file.te
+++ b/common/private/file.te
@@ -1,3 +1,5 @@
 type sdcard_posix, sdcard_type, sdcard_posix_contextmount_type, fs_type, mlstrustedobject;
-type adbroot_data_file, file_type, data_file_type, core_data_file_type;
 type sysfs_perdev_minors, fs_type, sysfs_type;
+type lineage_platform_framework_file, file_type, system_file_type;
+type lineage_permission_file, file_type, system_file_type;
+type lineage_privapp_pkg_file, file_type, system_file_type;
diff --git a/common/private/file_contexts b/common/private/file_contexts
index d4c34e9..dcca52f 100644
--- a/common/private/file_contexts
+++ b/common/private/file_contexts
@@ -7,6 +7,52 @@
 # OTA packages
 /data/lineageos_updates(/.*)?           u:object_r:ota_package_file:s0
 
+# Keep system_ext policy blob out of untrusted app reach.
+/(system_ext|system/system_ext)/etc/selinux/system_ext_sepolicy\.cil u:object_r:sepolicy_file:s0
+
+# Keep Lineage platform framework artifacts out of untrusted app reach.
+/system/framework/org\.system\.platform-res\.apk u:object_r:lineage_platform_framework_file:s0
+/system/framework/org\.lineageos\.platform\.jar u:object_r:lineage_platform_framework_file:s0
+/system/framework/oat/arm64/org\.lineageos\.platform\.vdex u:object_r:lineage_platform_framework_file:s0
+/system/framework/oat/arm64/org\.lineageos\.platform\.odex u:object_r:lineage_platform_framework_file:s0
+
+# Keep Lineage feature permission XML out of untrusted app reach.
+/(product|system/product)/etc/permissions/org\.lineageos\.settings\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.trust\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.profiles\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.android\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.hardware\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.globalactions\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.livedisplay\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/permissions/org\.lineageos\.health\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/permissions/LineageParts\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/permissions/LineageSettingsProvider\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/permissions/LineageSetupWizard\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/permissions/privapp_whitelist_co\.aospa\.sense\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/priv-app/LineageParts(/.*)? u:object_r:lineage_privapp_pkg_file:s0
+/(system_ext|system/system_ext)/priv-app/LineageSettingsProvider(/.*)? u:object_r:lineage_privapp_pkg_file:s0
+/(system_ext|system/system_ext)/priv-app/LineageSetupWizard(/.*)? u:object_r:lineage_privapp_pkg_file:s0
+/(system_ext|system/system_ext)/etc/default-permissions/default-permissions_org\.lineageos\.profiles\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/default-permissions/default-permissions_org\.lineageos\.updater\.xml u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/init/init\.lineage-system_ext\.rc u:object_r:lineage_permission_file:s0
+/(system_ext|system/system_ext)/etc/init/init\.lineage-updater\.rc u:object_r:lineage_permission_file:s0
+/(vendor|system/vendor)/etc/init/android\.hardware\.ir-service\.lineage\.rc u:object_r:lineage_permission_file:s0
+/(vendor|system/vendor)/etc/init/vendor\.lineage\.health-service\.default\.rc u:object_r:lineage_permission_file:s0
+/(vendor|system/vendor)/etc/init/vendor\.lineage\.touch-service\.xiaomi\.rc u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/initial-package-stopped-states-org\.lineageos\.aperture\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/initial-package-stopped-states-org\.lineageos\.etar\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/initial-package-stopped-states-org\.lineageos\.glimpse\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/initial-package-stopped-states-org\.lineageos\.jelly\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/initial-package-stopped-states-org\.lineageos\.twelve\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/lineage-component-overrides\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/lineage-sysconfig\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/org\.lineageos\.etar\.allowlist\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/preinstalled-packages-org\.lineageos\.aperture\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/preinstalled-packages-org\.lineageos\.etar\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/preinstalled-packages-org\.lineageos\.glimpse\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/preinstalled-packages-org\.lineageos\.jelly\.xml u:object_r:lineage_permission_file:s0
+/(product|system/product)/etc/sysconfig/preinstalled-packages-org\.lineageos\.twelve\.xml u:object_r:lineage_permission_file:s0
+
 # Postinstall
 /system/bin/backuptool_ab\.functions              u:object_r:postinstall_exec:s0
 /system/bin/backuptool_ab\.sh                     u:object_r:postinstall_exec:s0
diff --git a/common/private/init.te b/common/private/init.te
new file mode 100644
index 0000000..145008f
--- /dev/null
+++ b/common/private/init.te
@@ -0,0 +1,2 @@
+# Allow init to parse relabeled init rc fragments under /system_ext/etc/init and /vendor/etc/init.
+allow init lineage_permission_file:file r_file_perms;
diff --git a/common/private/lineage_platform_framework.te b/common/private/lineage_platform_framework.te
new file mode 100644
index 0000000..c40d0e9
--- /dev/null
+++ b/common/private/lineage_platform_framework.te
@@ -0,0 +1,2 @@
+allow { appdomain coredomain -untrusted_app_all } lineage_platform_framework_file:file r_file_perms;
+allow zygote lineage_platform_framework_file:file { execute read open getattr map };
diff --git a/common/private/lineage_privapp_pkg.te b/common/private/lineage_privapp_pkg.te
new file mode 100644
index 0000000..d04a987
--- /dev/null
+++ b/common/private/lineage_privapp_pkg.te
@@ -0,0 +1,9 @@
+# Keep selected /system_ext/priv-app Lineage packages out of untrusted app reach.
+allow system_server lineage_privapp_pkg_file:dir r_dir_perms;
+allow system_server lineage_privapp_pkg_file:file r_file_perms;
+allow system_server lineage_privapp_pkg_file:lnk_file r_file_perms;
+
+# Trusted app domains that may still need visibility.
+allow { platform_app system_app priv_app } lineage_privapp_pkg_file:dir r_dir_perms;
+allow { platform_app system_app priv_app } lineage_privapp_pkg_file:file r_file_perms;
+allow { platform_app system_app priv_app } lineage_privapp_pkg_file:lnk_file r_file_perms;
diff --git a/common/private/lineage_visibility_access.te b/common/private/lineage_visibility_access.te
new file mode 100644
index 0000000..697eb11
--- /dev/null
+++ b/common/private/lineage_visibility_access.te
@@ -0,0 +1,3 @@
+# Allow trusted app domains to read relabeled Lineage permission XMLs if needed.
+allow platform_app lineage_permission_file:file r_file_perms;
+allow system_app lineage_permission_file:file r_file_perms;
diff --git a/common/private/system_server.te b/common/private/system_server.te
index fdac25c..9768f36 100644
--- a/common/private/system_server.te
+++ b/common/private/system_server.te
@@ -1,6 +1,7 @@
 allow system_server storage_stub_file:dir getattr;
 
-allow system_server adbroot_service:service_manager find;
+# Allow SystemConfig/permission parsing to read relabeled Lineage permission XMLs.
+allow system_server lineage_permission_file:file r_file_perms;
 
 # Use HALs
 hal_client_domain(system_server, hal_lineage_health)
@@ -13,3 +14,6 @@ get_prop(system_server, vendor_persist_camera_prop)
 
 # Let system server toggle xtra-daemon
 set_prop(system_server, xtra_control_prop)
+
+# Let LineageSettingsProvider bump sys.lineage_settings_*_version props.
+set_prop(system_server, custom_version_prop)
diff --git a/common/private/untrusted_app.te b/common/private/untrusted_app.te
index 77959f6..0a61ec6 100644
--- a/common/private/untrusted_app.te
+++ b/common/private/untrusted_app.te
@@ -1 +1,19 @@
-allow untrusted_app_all lineage_profile_service:service_manager find;
+neverallow untrusted_app_all lineage_profile_service:service_manager find;
+
+# Untrusted apps must not read policy blobs.
+neverallow untrusted_app_all sepolicy_file:file r_file_perms;
+neverallow untrusted_app_all file_contexts_file:file r_file_perms;
+neverallow untrusted_app_all lineage_platform_framework_file:file rx_file_perms;
+neverallow untrusted_app_all lineage_permission_file:file r_file_perms;
+neverallow untrusted_app_all lineage_privapp_pkg_file:dir r_dir_perms;
+neverallow untrusted_app_all lineage_privapp_pkg_file:file r_file_perms;
+neverallow untrusted_app_all lineage_privapp_pkg_file:lnk_file r_file_perms;
+neverallow isolated_app_all lineage_privapp_pkg_file:dir r_dir_perms;
+neverallow isolated_app_all lineage_privapp_pkg_file:file r_file_perms;
+neverallow isolated_app_all lineage_privapp_pkg_file:lnk_file r_file_perms;
+neverallow ephemeral_app lineage_privapp_pkg_file:dir r_dir_perms;
+neverallow ephemeral_app lineage_privapp_pkg_file:file r_file_perms;
+neverallow ephemeral_app lineage_privapp_pkg_file:lnk_file r_file_perms;
+neverallow sdk_sandbox_all lineage_privapp_pkg_file:dir r_dir_perms;
+neverallow sdk_sandbox_all lineage_privapp_pkg_file:file r_file_perms;
+neverallow sdk_sandbox_all lineage_privapp_pkg_file:lnk_file r_file_perms;
diff --git a/common/vendor/file_contexts b/common/vendor/file_contexts
index 2746446..3cf088c 100644
--- a/common/vendor/file_contexts
+++ b/common/vendor/file_contexts
@@ -1,6 +1,10 @@
 # Camera HAL
 /(vendor|system/vendor)/bin/hw/android\.hardware\.camera\.provider-service(_32)?\.lineage u:object_r:hal_camera_default_exec:s0
 
+# Keep vendor policy blob out of untrusted app reach.
+/(vendor|system/vendor)/etc/selinux/vendor_sepolicy\.cil u:object_r:sepolicy_file:s0
+/(vendor|system/vendor)/etc/selinux/vendor_file_contexts u:object_r:file_contexts_file:s0
+
 # Fingerprint HAL
 /(vendor|system/vendor)/bin/hw/android\.hardware\.biometrics\.fingerprint@2\.0-service u:object_r:hal_fingerprint_default_exec:s0
 /(vendor|system/vendor)/bin/hw/android\.hardware\.biometrics\.fingerprint-service\.lineage u:object_r:hal_fingerprint_default_exec:s0
diff --git a/common/vendor/hal_lineage_health_default.te b/common/vendor/hal_lineage_health_default.te
index d3b7c16..c978d0b 100644
--- a/common/vendor/hal_lineage_health_default.te
+++ b/common/vendor/hal_lineage_health_default.te
@@ -5,3 +5,5 @@ type hal_lineage_health_default_exec, exec_type, vendor_file_type, file_type;
 init_daemon_domain(hal_lineage_health_default)
 
 binder_call(hal_lineage_health_default, servicemanager)
+
+allow hal_lineage_health_default self:netlink_kobject_uevent_socket create_socket_perms_no_ioctl;
diff --git a/qcom/vendor/hal_perf_default.te b/qcom/vendor/hal_perf_default.te
index 9125357..1038217 100644
--- a/qcom/vendor/hal_perf_default.te
+++ b/qcom/vendor/hal_perf_default.te
@@ -1,3 +1,4 @@
 r_dir_file(hal_perf_default, hal_power_default)
+rw_dir_file(hal_perf_default, proc_sched)
 
 rw_dir_file(hal_perf_default, proc_sched)
-- 
2.43.0

